### PR TITLE
[SPARK-38759][PYTHON][SS] Add StreamingQueryListener support in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR adds the support of `StreamingQueryListener` in PySpark. This PR adds the Python classes corresponding to Scala classes below:
- `StreamingQueryListener`
- `QueryStartedEvent`
- `QueryProgressEvent`
- `QueryTerminatedEvent`
- `StreamingQueryProgress`
- `StateOperatorProgress`
- `SourceProgress`
- `SinkProgress`

with the listener API, `StreamingQueryManager.addListener`

### Why are the changes needed?

This is a blocker of SPARK-38760. `DataFrame.observe` is an imporant feature for SS users to monitor their queries. However, this is missing in PySpark.

### Does this PR introduce _any_ user-facing change?

Yes, it adds a new API and class as described above. Users can do:

```python
from pyspark.sql.streaming import StreamingQueryListener


class Listener(StreamingQueryListener):
    def onQueryStarted(self, event: "QueryStartedEvent") -> None:
        print("Wow!")
    def onQueryProgress(self, event: "QueryProgressEvent") -> None:
        print("Hello!")
    def onQueryTerminated(self, event: "QueryTerminatedEvent") -> None:
        print("HI!")


spark.streams.addListener(Listener())


rates = spark.readStream.format("rate").option("rowsPerSecond", 1).load()
query = rates.writeStream.format("console").start()
query.awaitTermination()
```

### How was this patch tested?

Unittests added

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @hvanhovell @zsxwing @HeartSaVioR @viirya FYI. I am going to focus on writing some tests but the main codes here are almost ready. Please take a look when you find some time 🙏

File: sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala
Comment (xinrong-databricks) [inline sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala:86]: May I ask how `toPyRow` relates to StreamingQueryListener support?

File: python/pyspark/sql/streaming/listener.py
Comment (xinrong-databricks) [inline python/pyspark/sql/streaming/listener.py:342]: I may not see how the docstring relates to `observedMetrics`.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala:117]: I can interpret this as "When fixing here, fix there as well." I agree having slightly more verbose comment may be better to clarify for future readers having no/less context.

File: python/pyspark/sql/streaming/listener.py
Comment (viirya) [inline python/pyspark/sql/streaming/listener.py:32]: `pyspark.streaming.StreamingQuery` or `pyspark.sql.streaming.StreamingQuery`?

File: python/pyspark/sql/streaming/listener.py
Comment (alex-balikov) [inline python/pyspark/sql/streaming/listener.py:436]: we probably want doc strings for all api methods/properties

**Comment by HyukjinKwon** [comment]
Hey, thanks for reviews guys. I will take a closer look tmr, and address them.

**Comment by HeartSaVioR** [comment]
I see review comments about the doc which seem to be just copied from Scala/Java API doc.

Since this PR focuses mainly to deal with feature parity, how about simply allowing copy-paste doc from Scala/Java API doc (with additional content if something only applies to PySpark), and having another PR fixing both Scala/Java and PySpark doc altogether after this has merged?

The doc for Scala/Java API has been served for years and personally I'd consider commenting on the doc as proposing changes on "existing doc" instead of new doc, since the doc change here should trigger the change on Scala/Java doc to be in sync, effectively making changes on existing doc.

Does it make sense for everyone?

**Comment by HyukjinKwon** [comment]
Yeah, actually that's what I was going to point out. Should be better to create a separate PR to improve the documentation for both sides :-).

File: python/pyspark/sql/streaming/listener.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/listener.py:27]: So I intend to expose only `StreamingQueryListener` as a public API. Other events are NOT exposed as an API in PySpark for now. The events and progresses here are only supposed to be accessed as an instance.

File: python/docs/source/reference/pyspark.ss.rst
Comment (HyukjinKwon) [inline python/docs/source/reference/pyspark.ss.rst:33]: This was removed because this isn't an API.

**Review by xinrong-meng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Thanks @xinrong-databricks. Will merge this one if there are no more reviews for few days.

**Comment by HyukjinKwon** [comment]
Thanks @xinrong-databricks.

Merged to master.

FYI: I will work on related items such as like `DataFram.observe`, `StreamingQueryManager.removeListener` and `StereamingQueryManager.listListeners` with fixing the codes and tests further. i will follow up and address post-hoc review comments if there are 👍

**Review by dongjoon-hyun — COMMENTED** [review]
What a nice addition! Thank you, @HyukjinKwon and all.

**Comment by HeartSaVioR** [comment]
@HyukjinKwon 
I realized we have a section for describing streaming query listener in guide doc which has examples in various languages. 
https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#reporting-metrics-programmatically-using-asynchronous-apis

With this PR we would like to update Python tab to have an identical example with Scala/Java.
Would you mind if I ask to update the guide doc as well? Thanks in advance!

**Comment by HyukjinKwon** [comment]
oh yeah. will take a look. thanks for pointing this out!

**Pull request closed without merging**
